pkg/migration-tool/writer/writer.go

```diff
@@ -120,12 +130,42 @@ func (w *Write) Run(errChan chan<- error) {
 					errChan <- fmt.Errorf("remote-write run: %w", err)
 					return
 				}
+				planner.PutSlab(slabRef)
+				w.collectGarbage()
 			}
 		}
 	}()
 }
 
-// Blocks returns the total number of blocks pushed to the remote-write storage.
+var gcRunner sync.Once
+
+func (w *Write) collectGarbage() {
+	if w.sigGC == nil {
+		// sigGC is nil if w.GarbageCollectOnPush is set to false.
+		return
+	}
+	gcRunner.Do(func() {
+		go gc(w.sigGC)
+	})
+	w.sigGC <- struct{}{}
```

Comment:
Make the send non-blocking. If the channel already has 2 items `gc` can still block by working in `runtime.GC()`. By making this a non-blocking send you avoid this scenario and still guarantee that the gc will run since the queue has items.
```suggestion
    select {
    case w.sigGC <- struct{}{}:
    default:
    }
```

Reply:
Oh yes, this is much better way